polygon/p2p/fetcher.go

```diff
@@ -54,36 +60,53 @@ func (f *fetcher) FetchHeaders(ctx context.Context, start uint64, end uint64, pe
 	}
 
 	amount := end - start
-	requestId := f.requestIdGenerator()
-	observer := make(ChanMessageObserver[*sentry.InboundMessage])
+	if amount > maxFetchHeadersRange {
+		return nil, &ErrInvalidFetchHeadersRange{
+			start: start,
+			end:   end,
+		}
+	}
+
+	// Soft response limits are:
+	//   1. 2 MB size
+	//   2. 1024 headers
+	//
+	// A header is approximately 500 bytes, hence 1024 headers should be less than 2 MB.
+	// As a simplification we can only use MaxHeadersServe for chunking.
+	chunks := amount / eth.MaxHeadersServe
+	if amount%eth.MaxHeadersServe > 0 {
+		chunks++
+	}
 
+	observer := make(ChanMessageObserver[*sentry.InboundMessage], chunks)
 	f.messageListener.RegisterBlockHeadersObserver(observer)
 	defer f.messageListener.UnregisterBlockHeadersObserver(observer)
 
-	//
-	// TODO 1) chunk request into smaller ranges if needed to fit in the 2 MiB response size soft limit
-	//      and also 1024 max headers server (check AnswerGetBlockHeadersQuery)
-	//      2) peer should return <= amount, check for > amount and penalize peer
-	//
-	err := f.messageSender.SendGetBlockHeaders(ctx, peerId, eth.GetBlockHeadersPacket66{
-		RequestId: requestId,
-		GetBlockHeadersPacket: &eth.GetBlockHeadersPacket{
-			Origin: eth.HashOrNumber{
-				Number: start,
+	requestIds := make(map[uint64]uint64, chunks)
+	for i := uint64(0); i < chunks; i++ {
+		chunkStart := start + i*eth.MaxHeadersServe
+		chunkAmount := mathutil.MinUint64(end-chunkStart, eth.MaxHeadersServe)
+		requestId := f.requestIdGenerator()
+		requestIds[requestId] = i
+
+		if err := f.messageSender.SendGetBlockHeaders(ctx, peerId, eth.GetBlockHeadersPacket66{
```

Comment:
we shouldn't spam a peer with requests. let's do it sequentially: send one, and wait for a response, then send a 2nd one. if the response doesn't come in time, we should retry the request a couple of times.

Reply:
ok, this actually simplifies the code a lot - fixed
re: retrying - im doing this in a subsequent PR https://github.com/ledgerwatch/erigon/pull/9537 (it needs to be updated with the changes ive just done here)